File: backend/app/schemas/accidents.py
# -*- coding: utf-8 -*-
from __future__ import annotations

from datetime import datetime
from typing import Any
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class AccidentParseCreateRequest(BaseModel):
    model_config = {"extra": "ignore"}

    input_mode: Literal["auto", "manual"] = "auto"
    message_raw: str | None = Field(default=None, description="최초보고 메시지 원문")
    template_name: str | None = None
    site_standard_name: str | None = None
    reporter_name: str | None = None
    accident_datetime_text: str | None = None
    accident_datetime: datetime | None = None
    accident_place: str | None = None
    work_content: str | None = None
    injured_person_name: str | None = None
    accident_circumstance: str | None = None
    accident_reason: str | None = None
    injured_part: str | None = None
    diagnosis_name: str | None = None
    action_taken: str | None = None
    status: str | None = None
    management_category: str | None = None
    notes: str | None = None
    parse_status_override: str | None = None
    parse_note_override: str | None = None

    @model_validator(mode="after")
    def validate_by_mode(self):
        if self.input_mode == "auto":
            if not (self.message_raw or "").strip():
                raise ValueError("자동입력 모드에서는 message_raw가 필요합니다.")
            return self

        required = {
            "현장명": self.site_standard_name,
            "보고자": self.reporter_name,
            "사고시각": self.accident_datetime_text or self.accident_datetime,
            "사고장소": self.accident_place,
            "작업내용": self.work_content,
        }
        missing = [label for label, value in required.items() if not value]
        if missing:
            raise ValueError(f"수동입력 필수값 누락: {', '.join(missing)}")
        return self


class AccidentUpdateRequest(BaseModel):
    model_config = {"extra": "ignore"}

    site_standard_name: str
    reporter_name: str | None = None
    status: str
    management_category: str
    accident_datetime_text: str | None = None
    accident_datetime: datetime | None = None
    accident_place: str | None = None
    work_content: str | None = None
    injured_person_name: str | None = None
    accident_circumstance: str | None = None
    accident_reason: str | None = None
    injured_part: str | None = None
    diagnosis_name: str | None = None
    action_taken: str | None = None
    notes: str | None = None
    initial_report_template: str | None = None


class AccidentAttachmentItem(BaseModel):
    id: int
    file_name: str
    stored_path: str
    content_type: str | None = None
    file_size: int | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AccidentLookups(BaseModel):
    statuses: list[str]
    management_categories: list[str]
    site_names: list[str]


class AccidentListItem(BaseModel):
    id: int
    accident_id: str
    display_code: str
    parse_status: str
    site_name: str | None
    site_standard_name: str | None
    injured_person_name: str | None
    accident_datetime_text: str | None
    accident_datetime: datetime | None
    status: str
    management_category: str
    is_complete: bool
    has_attachments: bool
    nas_folder_path: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AccidentDetail(BaseModel):
    id: int
    accident_id: str
    display_code: str
    report_type: str
    source_type: str
    message_raw: str
    site_name: str | None
    reporter_name: str | None
    accident_datetime_text: str | None
    accident_datetime: datetime | None
    accident_place: str | None
    work_content: str | None
    injured_person_name: str | None
    accident_circumstance: str | None
    accident_reason: str | None
    injured_part: str | None
    diagnosis_name: str | None
    action_taken: str | None
    status: str
    management_category: str
    site_standard_name: str | None
    initial_report_template: str | None
    is_complete: bool
    nas_folder_path: str | None
    nas_folder_key: str | None
    notes: str | None
    parse_status: str
    parse_note: str | None
    created_by_user_id: int | None
    updated_by_user_id: int | None
    created_at: datetime
    updated_at: datetime
    attachments: list[AccidentAttachmentItem] = []

    model_config = {"from_attributes": True}


class AccidentInitialReportOutput(BaseModel):
    accident_id: int
    accident_code: str
    display_code: str
    parse_status: str
    composed_line: str
    fields: dict[str, Any]
    message_raw: str


class AccidentImportResult(BaseModel):
    imported: int
    skipped: int


class AccidentMasterExcelSyncResult(BaseModel):
    target_path: str
    backup_path: str | None
    exported_count: int
    excluded_count: int
    backup_created: bool


class AccidentParsePreviewResponse(BaseModel):
    parse_status: str
    parse_note: str | None
    composed_line: str
    fields: dict[str, Any]
    message_raw: str


class AccidentWorklistSection(BaseModel):
    count: int
    items: list[AccidentListItem]


class AccidentWorklistResponse(BaseModel):
    unverified: AccidentWorklistSection
    parse_review: AccidentWorklistSection
    missing_attachments: AccidentWorklistSection
    recent: AccidentWorklistSection

File: frontend/src/services/accidents.ts
import { api } from "@/services/api";

export interface AccidentAttachmentItem {
  id: number;
  file_name: string;
  stored_path: string;
  content_type: string | null;
  file_size: number | null;
  created_at: string;
}

export interface AccidentListItem {
  id: number;
  accident_id: string;
  display_code: string;
  parse_status: string;
  site_name: string | null;
  site_standard_name: string | null;
  injured_person_name: string | null;
  contractor_name: string | null;
  construction_team: string | null;
  accident_type: string | null;
  accident_datetime_text: string | null;
  accident_datetime: string | null;
  status: string;
  management_category: string;
  industrial_accident_report_status: string;
  medical_opinion_status: string;
  nas_sync_status: string;
  is_complete: boolean;
  has_attachments: boolean;
  nas_folder_path: string | null;
  created_at: string;
  updated_at: string;
}

export interface AccidentDetail extends AccidentListItem {
  report_type: string;
  source_type: string;
  message_raw: string;
  reporter_name: string | null;
  injured_person_birth_date: string | null;
  site_manager_name: string | null;
  work_team: string | null;
  job_name: string | null;
  injury_severity: string | null;
  leave_period: string | null;
  industrial_accident_status: string | null;
  accident_place: string | null;
  work_content: string | null;
  accident_circumstance: string | null;
  accident_reason: string | null;
  injured_part: string | null;
  diagnosis_name: string | null;
  action_taken: string | null;
  initial_report_template: string | null;
  nas_folder_key: string | null;
  notes: string | null;
  industrial_accident_report_submitted_at: string | null;
  medical_opinion_received_at: string | null;
  medical_opinion_summary: string | null;
  nas_sync_version: number;
  nas_synced_version: number;
  nas_synced_at: string | null;
  nas_sync_error: string | null;
  parse_note: string | null;
  created_by_user_id: number | null;
  updated_by_user_id: number | null;
  updated_at: string;
  attachments: AccidentAttachmentItem[];
  composed_line?: string | null;
  output_fields?: Record<string, unknown> | null;
}

export interface AccidentSyncResponse {
  server_time: string;
  upserts: AccidentListItem[];
}

export interface AccidentInitialReportOutput {
  accident_id: number;
  accident_code: string;
  display_code: string;
  parse_status: string;
  composed_line: string;
  fields: Record<string, unknown>;
  message_raw: string;
}

export interface AccidentLookups {
  statuses: string[];
  management_categories: string[];
  site_names: string[];
  industrial_accident_report_statuses: string[];
  medical_opinion_statuses: string[];
}

export interface AccidentUpdatePayload {
  site_standard_name: string;
  reporter_name: string | null;
  status: string;
  management_category: string;
  accident_datetime_text: string | null;
  accident_datetime: string | null;
  accident_place: string | null;
  work_content: string | null;
  injured_person_name: string | null;
  injured_person_birth_date: string | null;
  contractor_name: string | null;
  construction_team: string | null;
  site_manager_name: string | null;
  work_team: string | null;
  job_name: string | null;
  accident_type: string | null;
  injury_severity: string | null;
  leave_period: string | null;
  industrial_accident_status: string | null;
  accident_circumstance: string | null;
  accident_reason: string | null;
  injured_part: string | null;
  diagnosis_name: string | null;
  action_taken: string | null;
  notes: string | null;
  initial_report_template: string | null;
  industrial_accident_report_status: string;
  industrial_accident_report_submitted_at: string | null;
  medical_opinion_status: string;
  medical_opinion_received_at: string | null;
  medical_opinion_summary: string | null;
}

export interface AccidentCreatePayload extends AccidentUpdatePayload {
  input_mode: "auto" | "manual";
  template_name?: string | null;
  message_raw: string | null;
  parse_status_override?: string | null;
  parse_note_override?: string | null;
}

export interface AccidentWorklistSection {
  count: number;
  items: AccidentListItem[];
}

export interface AccidentWorklistResponse {
  unverified: AccidentWorklistSection;
  parse_review: AccidentWorklistSection;
  missing_attachments: AccidentWorklistSection;
  recent: AccidentWorklistSection;
}

export interface AccidentParsePreviewResponse {
  parse_status: string;
  parse_note: string | null;
  composed_line: string;
  fields: Record<string, unknown>;
  message_raw: string;
}

export interface AccidentMasterExcelSyncResult {
  target_path: string;
  backup_path: string | null;
  exported_count: number;
  excluded_count: number;
  backup_created: boolean;
}

export async function fetchAccidentLookups() {
  const res = await api.get<AccidentLookups>("/accidents/lookups");
  return res.data;
}

export async function fetchAccidents(params: {
  queues?: string[];
  show_all?: boolean;
}) {
  const res = await api.get<AccidentListItem[]>("/accidents", { params });
  return res.data ?? [];
}

export async function syncAccidents(since?: string | null) {
  const res = await api.get<AccidentSyncResponse>("/accidents/sync", {
    params: since ? { since } : undefined,
  });
  return res.data;
}

export async function fetchAccidentDetail(id: number) {
  const res = await api.get<AccidentDetail>(`/accidents/${id}`);
  return res.data;
}

export async function fetchAccidentInitialReport(id: number) {
  const res = await api.get<AccidentInitialReportOutput>(`/accidents/${id}/initial-report`);
  return res.data;
}

export async function fetchAccidentWorklist() {
  const res = await api.get<AccidentWorklistResponse>("/accidents/worklist");
  return res.data;
}

export async function fetchAccidentParsePreview(payload: { message_raw: string }) {
  const res = await api.post<AccidentParsePreviewResponse>("/accidents/initial-report/parse-preview", payload);
  return res.data;
}

export async function syncAccidentsToMasterExcel() {
  const res = await api.post<AccidentMasterExcelSyncResult>("/accidents/export-to-master-excel");
  return res.data;
}

export async function downloadAccidentMaster(downloadPassword: string): Promise<Blob> {
  const res = await api.post(
    "/accidents/export/master",
    { download_password: downloadPassword },
    { responseType: "blob" },
  );
  return new Blob([res.data], {
    type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  });
}

/** 백엔드 검증 후 탐색기 실행용 .bat 파일을 내려받는다(Windows에서 더블클릭). */
export async function downloadAccidentNasFolderLauncher(accidentPk: number): Promise<void> {
  const res = await api.get(`/accidents/${accidentPk}/nas-folder-launcher`, { responseType: "blob" });
  const blob = new Blob([res.data], { type: "application/octet-stream" });
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `open_accident_${accidentPk}_nas.bat`;
  a.rel = "noopener";
  a.click();
  window.URL.revokeObjectURL(url);
}
